### PR TITLE
Cleanup evar_extra: remove evar_info's store and add maps to evar_map

### PR DESCRIPTION
**Pull request by mattam82**
**Kind:** cleanup

This PR is a replacement for PR #611, making the handling of evar_flags (obligation, evar resolvability and restricted flags) centralized in the evar_map using evar sets and maps in a record instead of each evar_info's evar_extra. It removes evar_extra which was only used for these purposes. Primitive are now exported in Evd to set/unset the resolability and obligation flag. The restricted flag is handled internally during evar restriction. Now all merging of these flags is done explicitly, making the code easier to follow, and we get rid of a dynamic store field for every evar. This breaks API compatibility lightly (the new_evar functions do not have an optional store argument anymore).

### CODE REVIEW

File: pretyping/pretyping.ml [hunk 1]
```diff
@@ -517,6 +517,15 @@ let rec pretype k0 resolve_tc (tycon : type_constraint) (env : GlobEnv.t) (sigma
         | Some ty -> sigma, ty
         | None -> new_type_evar env sigma loc in
       let sigma, uj_val = new_evar env sigma ~src:(loc,k) ~naming ty in
+      let sigma =
```
Comment (gares) [inline pretyping/pretyping.ml]: How is this change related to this PR?

File: proofs/goal.ml [hunk 2]
```diff
@@ -67,11 +62,10 @@ module V82 = struct
 		Evd.evar_filter = Evd.Filter.identity;
 		Evd.evar_body = Evd.Evar_empty;
 		Evd.evar_source = (Loc.tag Evar_kinds.GoalEvar);
-		Evd.evar_candidates = None;
-		Evd.evar_extra = extra }
+                Evd.evar_candidates = None }
```
Comment (gares) [inline proofs/goal.ml]: indentation seems weird

**Review by gares — APPROVED** [review]
The change makes sense to me

File: engine/evd.ml [hunk 3]
```diff
@@ -425,6 +427,7 @@ type evar_map = {
   last_mods  : Evar.Set.t;
   (** Metas *)
   metas      : clbinding Metamap.t;
+  evar_flags : evar_flags;
```
Comment (ppedrot) [inline engine/evd.ml]: Rather than a set at the level of evar maps, I think it is probably slightly better to use a flag in the evarinfo. Depending on the kind of uses, this would be more efficient. Also, you could pack the obligation and unresolvable flags into only one integer field. Allocation of evar-related data is quite sensitive, you can easily make compcert or VST go crazy if you're not careful.

File: pretyping/pretyping.ml (hunk 1 above)
Comment (mattam82) [reply pretyping/pretyping.ml]: It moves the handling of the obligation flag higher in the hierarchy, avoiding to do strange comparisons of the evar kind in unification.

File: proofs/goal.ml (hunk 2 above)
Comment (mattam82) [reply proofs/goal.ml]: Right.

Comment (mattam82) [reply proofs/goal.ml]: That's because the new line doesn't use tabs, shall I untabify the rest?

**Review by ejgallego — COMMENTED** [review]
I am not an expert but it looks very good!

File: engine/evd.ml (hunk 3 above)
Comment (mattam82) [reply engine/evd.ml]: Using this solution avoids polluting the evar_infos of everything. For unresolvability the next step is to actually replace the evar set of unresolvables with a set of typeclass evars to be resolved (setting it to empty to ensure nested resolutions don't fire the same resolution again). For restricted you might be right, it could help performance. I'd like to see numbers for this solution first.

Comment (mattam82) [reply engine/evd.ml]: My understanding of the VST slowdown was that we were testing if a big evar map of undefined evars had resolvable ones, which will now result in a lookup in the unresolvable_evars set (which should contain all the goals) instead of testing an evar_extra field. So this could definitely have some impact. It might require inverting the logic to record only the resolvable goals, which should be a small set in general.

File: proofs/goal.ml (hunk 2 above)
Comment (gares) [reply proofs/goal.ml]: as you prefer

**Comment by mattam82** [comment]
I’ll make overlays and launch a bench.

**Comment by mattam82** [comment]
Benchmarking here: https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/545/

File: engine/evd.ml (hunk 3 above)
Comment (mattam82) [reply engine/evd.ml]: Indeed, the bench shows a large impact on vst, I'll refine the proposal.

**Comment by ppedrot** [comment]
As forecast, this PR has fairly dire consequences on VST:
```
┌──────────────────────────┬─────────────────────────┬───────────────────────────────────────┬────────────────────────────────────────┬─────────────────────────┬──────────────────────┐
│                          │      user time [s]      │              CPU cycles               │            CPU instructions            │  max resident mem [KB]  │      mem faults      │
│                          │                         │                                       │                                        │                         │                      │
│             package_name │     NEW     OLD PDIFF   │            NEW            OLD PDIFF   │            NEW            OLD  PDIFF   │     NEW     OLD PDIFF   │  NEW  OLD    PDIFF   │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│               coq-sf-vfa │   47.10   47.62 -1.09 % │   130732794015   131700565421 -0.73 % │   209494782429   209702394531  -0.10 % │  534380  533436 +0.18 % │   28    6  +366.67 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│    coq-mathcomp-fingroup │   84.68   85.10 -0.49 % │   234283874456   235390298895 -0.47 % │   348282133999   348542846497  -0.07 % │  586392  587704 -0.22 % │    0  162  -100.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│   coq-mathcomp-character │  293.78  295.18 -0.47 % │   816210122889   820256015577 -0.49 % │  1189668988326  1191162079628  -0.13 % │ 1109728 1118488 -0.78 % │    1  106   -99.06 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│     coq-mathcomp-algebra │  205.26  206.01 -0.36 % │   569154686601   570831687821 -0.29 % │   798820602360   798847959823  -0.00 % │  643396  640628 +0.43 % │  312   13 +2300.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│    coq-mathcomp-solvable │  220.27  221.04 -0.35 % │   611737707922   613557167932 -0.30 % │   874581135904   875314049490  -0.08 % │  807260  857132 -5.82 % │   20    6  +233.33 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│       coq-mathcomp-field │  468.36  469.90 -0.33 % │  1303091346215  1307717253989 -0.35 % │  2127085871584  2125595379648  +0.07 % │  799060  804704 -0.70 % │    0    4  -100.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│      coq-formal-topology │   57.92   58.11 -0.33 % │   158107393599   158718446658 -0.38 % │   239490056336   239733358338  -0.10 % │  477680  477612 +0.01 % │    0    0     +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│   coq-mathcomp-odd-order │ 1417.41 1421.84 -0.31 % │  3949464817889  3959907035547 -0.26 % │  6658488223706  6658683627181  -0.00 % │ 1354348 1362736 -0.62 % │    9  210   -95.71 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│                 coq-corn │ 1542.82 1546.89 -0.26 % │  4280415678686  4291864272144 -0.27 % │  6350629646364  6347247544183  +0.05 % │  828188  818052 +1.24 % │   44   44    +0.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│   coq-mathcomp-ssreflect │   65.44   65.58 -0.21 % │   179830436071   180801708877 -0.54 % │   256345860146   256483264908  -0.05 % │  531812  532264 -0.08 % │   17  108   -84.26 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│          coq-lambda-rust │ 2322.42 2326.79 -0.19 % │  6469947173416  6481410298153 -0.18 % │  8961606181139  8969167848925  -0.08 % │ 1448792 1474204 -1.72 % │    0   17  -100.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│               coq-sf-plf │   74.77   74.89 -0.16 % │   206281878962   206702028700 -0.20 % │   300890547812   300656216088  +0.08 % │  519452  519576 -0.02 % │   15    0     +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│         coq-math-classes │  248.99  249.36 -0.15 % │   685842408053   686720151242 -0.13 % │   900087647593   899432929679  +0.07 % │  521872  521864 +0.00 % │  337   82  +310.98 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│         coq-fiat-parsers │  718.34  719.32 -0.14 % │  1991916206029  1994629293629 -0.14 % │  3058407972617  3056872640715  +0.05 % │ 2881888 2906472 -0.85 % │  525  757   -30.65 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│              coq-unimath │ 1330.73 1332.28 -0.12 % │  3692697711067  3697388133656 -0.13 % │  6223220342547  6222641146044  +0.01 % │  985376 1043328 -5.55 % │  342   19 +1700.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│               coq-geocoq │ 2841.26 2844.49 -0.11 % │  7897900453746  7903130809603 -0.07 % │ 12597948267867 12579857354980  +0.14 % │ 1290632 1290312 +0.02 % │  597  190  +214.21 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│                coq-flocq │  591.20  591.20 +0.00 % │  1644072332361  1642578835991 +0.09 % │  2406275116681  2403062272774  +0.13 % │ 1764796 1766800 -0.11 % │  452  366   +23.50 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│             coq-compcert │  891.96  891.18 +0.09 % │  2474444276199  2470324254506 +0.17 % │  3574561270777  3560554641849  +0.39 % │ 1347400 1349144 -0.13 % │  286  452   -36.73 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│                coq-sf-lf │   44.38   44.34 +0.09 % │   122445693940   122268818930 +0.14 % │   198968602357   199223943406  -0.13 % │  423160  422916 +0.06 % │  116    0     +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│ coq-mathcomp-real-closed │  188.52  188.35 +0.09 % │   524909914904   523909931776 +0.19 % │   791251299716   790158402967  +0.14 % │  827576  830576 -0.36 % │    2   13   -84.62 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│                 coq-hott │  340.80  340.37 +0.13 % │   943918781304   943953946036 -0.00 % │  1441109945941  1439986748252  +0.08 % │  666912  666928 -0.00 % │  432    0     +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│                coq-color │  735.93  734.89 +0.14 % │  2039815487376  2037030085101 +0.14 % │  2469823174813  2468740990953  +0.04 % │ 1370536 1370120 +0.03 % │   65  115   -43.48 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│          coq-fiat-crypto │ 6395.07 6384.68 +0.16 % │ 17785516167937 17755424876247 +0.17 % │ 29425870212688 29360034461495  +0.22 % │ 2515972 2560752 -1.75 % │ 1774 1352   +31.21 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│           coq-coquelicot │  104.08  103.90 +0.17 % │   287633974274   287230882096 +0.14 % │   386796885515   386694772809  +0.03 % │  713936  713024 +0.13 % │  281   13 +2061.54 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│              coq-bignums │   99.26   98.72 +0.55 % │   274122113336   273636236586 +0.18 % │   393795061081   393557951888  +0.06 % │  535052  539880 -0.89 % │  175   43  +306.98 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│            coq-fiat-core │  135.64  134.74 +0.67 % │   377237732351   376515798479 +0.19 % │   503979559406   503950036714  +0.01 % │  496516  497644 -0.23 % │  498  188  +164.89 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼────────────────────────────────────────┼─────────────────────────┼──────────────────────┤
│                  coq-vst │ 1864.75 1710.66 +9.01 % │  5185956686064  4756042476716 +9.04 % │  7265580582972  6522730468429 +11.39 % │ 2274196 2224152 +2.25 % │  495  837   -40.86 % │
└──────────────────────────┴─────────────────────────┴───────────────────────────────────────┴────────────────────────────────────────┴─────────────────────────┴──────────────────────┘
```

**Comment by ppedrot** [comment]
@mattam82 If you push an up-to-date fixed version of this PR I can merge it so that you can rebase #8741 on it. Otherwise we can simply wait for #8741 to land alone and close this PR.

**Comment by mattam82** [comment]
This part didn’t change. To solve the efficiency issue you need the #8741
commit. So I’ll just close this one.
